Wikipedia/Code/WMFAppViewController.m

```diff
@@ -1587,16 +1615,22 @@ - (void)checkRemoteAppConfigIfNecessary {
     self.checkingRemoteConfig = YES;
     CFAbsoluteTime lastCheckTime = (CFAbsoluteTime)[[self.dataStore.viewContext wmf_numberValueForKey:WMFLastRemoteAppConfigCheckAbsoluteTimeKey] doubleValue];
     CFAbsoluteTime now = CFAbsoluteTimeGetCurrent();
-    if (now - lastCheckTime >= WMFRemoteAppConfigCheckInterval) {
-        [self.dataStore updateLocalConfigurationFromRemoteConfigurationWithCompletion:^(NSError *error) {
-            if (!error) {
-                [self.dataStore.viewContext wmf_setValue:[NSNumber numberWithDouble:now] forKey:WMFLastRemoteAppConfigCheckAbsoluteTimeKey];
-            }
-            self.checkingRemoteConfig = NO;
-        }];
-    } else {
+    BOOL shouldCheckRemoteConfig = now - lastCheckTime >= WMFRemoteAppConfigCheckInterval || self.dataStore.failedToUpdateLocalConfig;
+    if (!shouldCheckRemoteConfig) {
         self.checkingRemoteConfig = NO;
+        return;
     }
+    self.dataStore.isLocalConfigUpdateAllowed = YES;
+    [self startRemoteConfigCheckBackgroundTask:^{
+        self.dataStore.isLocalConfigUpdateAllowed = NO;
+        [self endRemoteConfigCheckBackgroundTask];
```

Comment:
@nambatee I think endRemoteConfigCheckBackgroundTask needs to be called in the updateLocalConfigurationFromRemoteConfigurationWithCompletion completion block. Instead here would be the place to cancel any tasks/operations because the background expiration has hit.

Reply:
Good point, I think it should be called in both places - per documentation (https://developer.apple.com/documentation/uikit/uiapplication/1623051-beginbackgroundtask), (in the handler) _Failure to end the task explicitly will result in the termination of the app_

Reply:
Maybe this is fine actually, you could need it here to prevent the app from terminating. But do you also need it in updateLocalConfigurationFromRemoteConfigurationWithCompletion? Seems like you'd need to flag the system your operation is complete rather than wait for expiration but I could be getting it wrong. I was looking at how SavedArticlesFetcher works, `beginBackgroundTaskWithName:@"SavedArticlesFetch" ` expiration handler cancels the tasks and endBackgroundTask is called elsewhere.

Reply:
@nambatee yeah, both places sounds good